### PR TITLE
test: Add CHID matching test

### PR DESCRIPTION
**Pull request by anonymix007**

### CODE REVIEW

**Comment by anonymix007** [comment]
@bluca CI failures look unrelated to this PR. Are they not?

File: src/test/test-chid-match.c [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
```
Comment (bluca) [inline src/test/test-chid-match.c]: extra new line

File: src/test/generate-hwids-section.py
Comment (bluca) [inline src/test/generate-hwids-section.py:1]: Add this to the list of linted files in `.github/workflows/linter.yml`

File: src/test/chid-match-util.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include "smbios.h"
+
```
Comment (bluca) [inline src/test/chid-match-util.h]: extra newline

File: src/test/chid-match-util.c [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "chid-match-util.h"
+#include "efi.h"
+
+#include "./chid.c"
```
Comment (bluca) [inline src/test/chid-match-util.c]: Don't include c sources, instead export the function needed and include the source files in the compilation unit in meson

File: src/test/chid.c
Comment (bluca) [inline src/test/chid.c:1]: is this symlink necessary?

Comment (anonymix007) [reply src/test/chid.c:1]: Yes, it absolutely is. 
It's something with the include paths. Without this, compiler will use headers from `src/boot/` which simply don't work in userspace.

File: src/boot/chid.c [hunk 4]
```diff
@@ -14,9 +14,10 @@
  * https://github.com/aarch64-laptops/edk2/blob/dtbloader-app/EmbeddedPkg/Application/ConfigTableLoader/CHID.c
  */
 
+#include "efi.h"
```
Comment (bluca) [inline src/boot/chid.c]: why is this moved?

File: src/test/chid.c
Comment (bluca) [reply src/test/chid.c:1]: We have some other tests though, like `src/boot/test-efi-string.c` so it should be possible in principle to refactor the file a bit so that it works

File: src/test/util.h
Comment (bluca) [inline src/test/util.h:1]: should be called efi-util.h if it's for efi

File: src/boot/chid.c (hunk 4 above)
Comment (anonymix007) [reply src/boot/chid.c]: Because it doesn't compile otherwise. The order matters here

File: src/test/util.h
Comment (anonymix007) [reply src/test/util.h:1]: But it can't be called efi-util.h because it's called util.h everywhere in src/boot/*.{c,h}

File: src/boot/chid.c (hunk 4 above)
Comment (bluca) [reply src/boot/chid.c]: Please add a comment explaining it then, as headers are sorted alphabetically usually

File: src/test/chid.c
Comment (anonymix007) [reply src/test/chid.c:1]: I guess that works too

File: src/fundamental/efi-fundamental.h [hunk 5]
```diff
@@ -1,7 +1,7 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 #pragma once
 
-#if !SD_BOOT
+#if !defined(SD_BOOT_EFI_H_INCLUDED)
```
Comment (bluca) [inline src/fundamental/efi-fundamental.h]: can the includes be rearranged so that this is not necessary?

Comment (anonymix007) [reply src/fundamental/efi-fundamental.h]: I don't think so, it's not about the order really. 
`efi.h` and `efi-fundamental.h` are both defining these structures (and this is not the only situation when this happens) and both are included in `chid.c`, so compilation fails with SD_BOOT=0.

I guess we can remove these definitions from `efi.h` and only keep them in `efi-fundamental.h`.

Comment (bluca) [reply src/fundamental/efi-fundamental.h]: yeah de-duplicating sounds fine

File: src/boot/efi.h [hunk 6]
```diff
@@ -6,8 +6,11 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#include "efi-fundamental.h"
 #include "macro-fundamental.h"
 
+#define SD_BOOT_EFI_H_INCLUDED
```
Comment (bluca) [inline src/boot/efi.h]: leftover

**Review by bluca — APPROVED** [review]

File: src/boot/util.h [hunk 7]
```diff
@@ -236,3 +239,13 @@ char16_t *get_extra_dir(const EFI_DEVICE_PATH *file_path);
 
 #define bswap_16(x) __builtin_bswap16(x)
 #define bswap_32(x) __builtin_bswap32(x)
+
+#else
+
+#include <stdlib.h>
+#include <string.h>
+
+#define xnew0(type, n) ((type *) calloc((n), sizeof(type)))
```
Comment (yuwata) [inline src/boot/util.h]: `xnew0()` should trigger assertion on failure.
```c
#include "alloc-util.h"
#define xnew0(type, n) ASSERT_PTR(new0(type, n))
```

File: src/boot/generate-hwids-section.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+# easier to import the file.
+sys.path.append(os.path.dirname(__file__) + '/../ukify')
+import ukify
+
+BYTES_PER_LINE = 16
+
+hwids = ukify.parse_hwid_dir(Path(sys.argv[1]))
+
+print(
+    """#include <stddef.h>
```
Comment (yuwata) [inline src/boot/generate-hwids-section.py]: Please add SPDX license header to the generated source file.

File: src/boot/chid.c [hunk 9]
```diff
@@ -57,10 +57,19 @@ typedef struct SmbiosInfo {
         char16_t *smbios_fields[_CHID_SMBIOS_FIELDS_MAX];
 } SmbiosInfo;
 
-static void smbios_info_populate(SmbiosInfo *ret_info) {
+/* Only exposed for testing purposes */
+void smbios_info_populate_internal(SmbiosInfo *ret_info, bool reset);
```
Comment (yuwata) [inline src/boot/chid.c]: Why not to declare this in chid.h ??

File: src/boot/chid.c [hunk 10]
```diff
@@ -74,6 +83,10 @@ static void smbios_info_populate(SmbiosInfo *ret_info) {
         ret_info->smbios_fields[CHID_SMBIOS_BASEBOARD_MANUFACTURER] = smbios_to_hashable_string(raw.baseboard_manufacturer);
 }
 
+static void smbios_info_populate(SmbiosInfo *ret_info) {
```
Comment (yuwata) [inline src/boot/chid.c]: Then, make this inline in the header.

File: src/boot/test-chid-match.c [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "efi.h"
+
+#include "chid.h"
+#include "smbios.h"
+
+#include "tests.h"
```
Comment (yuwata) [inline src/boot/test-chid-match.c]: Please drop empty lines.

File: src/boot/test-chid-match.c [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+static RawSmbiosInfo current_info = {};
+
+/* This is a dummy implementation for testing purposes */
+void smbios_raw_info_populate(RawSmbiosInfo *ret_info) {
+        assert(ret_info);
+        *ret_info = current_info;
+}
+
+extern void smbios_info_populate_internal(void *ret_info, bool reset);
```
Comment (yuwata) [inline src/boot/test-chid-match.c]: Then, please drop this.

File: src/boot/test-chid-match.c [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+TEST(chid_match) {
+        for (size_t i = 0; i < ELEMENTSOF(smbios_info); i++) {
+                /* Reset cached SMBIOS strings and update for the current test */
+                smbios_info_populate_internal(NULL, true);
```
Comment (yuwata) [inline src/boot/test-chid-match.c]: Please add short comment to the boolean argument: `/* reset = */ true`

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/boot/chid.c (hunk 9 above)
Comment (anonymix007) [reply src/boot/chid.c]: Because it shouldn't be used for anything except tests.

There are the following approaches:
1. Declare it only where used (which is what I did)
2. Declare it as `void smbios_info_populate_internal(void *ret_info, bool reset)` with a comment like "do NOT use anywhere except for testing"
3. Drop caching altogether (will make things a bit slower if `chid_match` is called multiple times, i.e. for add-ons or different device types once those will be added).

I'm not sure what's the best approach here. What do you think should be chosen?

File: src/boot/chid.c (hunk 10 above)
Comment (anonymix007) [reply src/boot/chid.c]: Once again, this shouldn't be exposed

File: src/boot/chid.c (hunk 9 above)
Comment (yuwata) [reply src/boot/chid.c]: Typically, we choose 2 at several places. So, if there is no special reasons to 'hide' the function, please declare it in the header with a short comment.

File: src/boot/util.h [hunk 14]
```diff
@@ -236,3 +239,14 @@ char16_t *get_extra_dir(const EFI_DEVICE_PATH *file_path);
...
 #define bswap_32(x) __builtin_bswap32(x)
+
+#else
+
+#include <stdlib.h>
+#include <string.h>
+#include "alloc-util.h"
+
+#define xnew0(type, n) ASSERT_PTR(new0(type, n))
+#define efi_guid_equal(a, b) (memcmp((a), (b), sizeof(EFI_GUID)) == 0)
```
Comment (yuwata) [inline src/boot/util.h]: `efi_guid_equal()` is defined in efi-fundamental.h . So, it is not necessary here?

File: src/boot/util.h [hunk 15]
```diff
@@ -236,3 +239,14 @@ char16_t *get_extra_dir(const EFI_DEVICE_PATH *file_path);
 
 #define bswap_16(x) __builtin_bswap16(x)
 #define bswap_32(x) __builtin_bswap32(x)
+
+#else
+
+#include <stdlib.h>
+#include <string.h>
+#include "alloc-util.h"
```
Comment (yuwata) [inline src/boot/util.h]: I guess `stdlib.h` and `string.h` are not necessary.
At least, please insert an empty line between the system headers and our header.

File: src/boot/chid.c [hunk 16]
```diff
@@ -57,10 +57,16 @@ typedef struct SmbiosInfo {
...
+static void smbios_info_populate(SmbiosInfo *ret_info, bool reset) {
         static RawSmbiosInfo raw = {};
         static bool raw_info_populated = false;
 
+        if (reset) {
+                raw_info_populated = false;
+                return;
+        }
+
+        assert(ret_info);
```
Comment (yuwata) [inline src/boot/chid.c]: Instead, how about to introduce `smbios_raw_info_get_cached()` or so, and drop the static variables in `smbios_info_populate()`, like https://github.com/yuwata/systemd/commit/ab0761367333be31bcd51ab7b16048f399ed863c ? Feel free to pick the commit into this PR.

Then, you can drop the additional `bool rest` argument, and `chid_reset_smbios_cache()`.

**Review by yuwata — CHANGES_REQUESTED** [review]

**Review by yuwata — APPROVED** [review]
LGTM.
